gossip/hub/handlers.go

```diff
@@ -0,0 +1,664 @@
+		sendHTTPError(w, status, err)
+		return
+	}
+
+	// Additional check, for consistency the handler must return an error for non-200 status
+	if status != http.StatusOK {
+		glog.Warningf("%s: %s handler non 200 without error: %d %v", a.info.hubPrefix, a.epPath, status, err)
+		sendHTTPError(w, http.StatusInternalServerError, fmt.Errorf("http handler misbehaved, status: %d", status))
+		return
+	}
+}
+
+// sourceCryptoInfo holds information needed to check signatures generated by a source.
+type sourceCryptoInfo struct {
+	pubKeyData []byte // DER-encoded public key
+	pubKey     crypto.PublicKey
+	hasher     crypto.Hash
+}
+
+// hubInfo holds information about a specific hub instance.
+type hubInfo struct {
+	// Instance-wide options
+	opts InstanceOptions
+
+	hubPrefix string
+	logID     int64
+	urlPrefix string
+	rpcClient trillian.TrillianLogClient
+	signer    *tcrypto.Signer
+	cryptoMap map[string]sourceCryptoInfo
+}
+
+// newHubInfo creates a new instance of hubInfo.
+func newHubInfo(logID int64, prefix string, rpcClient trillian.TrillianLogClient, signer crypto.Signer, cryptoMap map[string]sourceCryptoInfo, opts InstanceOptions) *hubInfo {
+	info := &hubInfo{
+		opts:      opts,
+		hubPrefix: fmt.Sprintf("%s{%d}", prefix, logID),
+		logID:     logID,
+		urlPrefix: prefix,
+		rpcClient: rpcClient,
+		signer:    tcrypto.NewSigner(logID, signer, crypto.SHA256),
+		cryptoMap: cryptoMap,
+	}
+	once.Do(func() { setupMetrics(opts.MetricFactory) })
+	knownHubs.Set(1.0, strconv.FormatInt(logID, 10))
+
+	return info
+}
+
+// Handlers returns a map from URL paths (with the given prefix) and AppHandler instances
+// to handle those entrypoints.
+func (h *hubInfo) Handlers(prefix string) PathHandlers {
+	if !strings.HasPrefix(prefix, "/") {
+		prefix = "/" + prefix
+	}
+	prefix = strings.TrimRight(prefix, "/")
+
+	// Bind the hubInfo instance to give an appHandler instance for each entrypoint.
+	return PathHandlers{
+		prefix + api.PathPrefix + api.AddSignedBlobPath:     AppHandler{info: h, handler: addSignedBlob, epPath: api.AddSignedBlobPath, method: http.MethodPost},
+		prefix + api.PathPrefix + api.GetSTHPath:            AppHandler{info: h, handler: getSTH, epPath: api.GetSTHPath, method: http.MethodGet},
+		prefix + api.PathPrefix + api.GetSTHConsistencyPath: AppHandler{info: h, handler: getSTHConsistency, epPath: api.GetSTHConsistencyPath, method: http.MethodGet},
+		prefix + api.PathPrefix + api.GetProofByHashPath:    AppHandler{info: h, handler: getProofByHash, epPath: api.GetProofByHashPath, method: http.MethodGet},
+		prefix + api.PathPrefix + api.GetEntriesPath:        AppHandler{info: h, handler: getEntries, epPath: api.GetEntriesPath, method: http.MethodGet},
+		prefix + api.PathPrefix + api.GetSourceKeysPath:     AppHandler{info: h, handler: getSourceKeys, epPath: api.GetSourceKeysPath, method: http.MethodGet},
+	}
+}
+
+func addSignedBlob(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
```

Comment:
All these methods are a bit lengthy.
Would it be possible to separate the parsing and serializing from the implementation?
For example, just move the core implementation into a separate struct with typed methods.

```go
func addSignedBlob(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
      var req api.AddSignedBlobRequest
      if err := json.Unmarshal(body, &req); err != nil {
		glog.V(1).Infof("%s: Failed to parse request body: %v", c.hubPrefix, err)
		return http.StatusBadRequest, fmt.Errorf("failed to parse add-signed-blob body: %v", err)
	}


       resp, err := imp.AddSignedBlob(ctx, req)

        w.Header().Set(contentTypeHeader, contentTypeJSON)
	jsonData, err := json.Marshal(&resp)
	if err != nil {
		return http.StatusInternalServerError, fmt.Errorf("failed to JSON-marshal add-signed-blob rsp: %s", err)
	}
 	if _, err = w.Write(jsonData); err != nil {
		return http.StatusInternalServerError, fmt.Errorf("failed to write add-signed-blob rsp: %s", err)
	}
 	glog.V(3).Infof("%s: AddSignedBlob <= SGT", c.hubPrefix)
	return http.StatusOK, nil
}
      




```

Reply:
Done.